bundles/org.openhab.binding.solarforecast/src/main/java/org/openhab/binding/solarforecast/internal/forecastsolar/ForecastSolarObject.java

```diff
@@ -0,0 +1,290 @@
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class ForecastSolarObject implements SolarForecast {
+    private static final double UNDEF = -1;
+
+    private final Logger logger = LoggerFactory.getLogger(ForecastSolarObject.class);
+    private final TreeMap<ZonedDateTime, Double> wattHourMap = new TreeMap<ZonedDateTime, Double>();
+    private final TreeMap<ZonedDateTime, Double> wattMap = new TreeMap<ZonedDateTime, Double>();
+    private final DateTimeFormatter dateInputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+
+    private ZoneId zone = ZoneId.systemDefault();
+    private Optional<String> rawData = Optional.empty();
+    private boolean valid = false;
+    private Instant expirationDateTime;
+
+    public ForecastSolarObject() {
+        expirationDateTime = Instant.now();
+    }
+
+    public ForecastSolarObject(String content, Instant expirationDate) {
+        expirationDateTime = expirationDate;
+        if (!content.isEmpty()) {
+            rawData = Optional.of(content);
+            try {
+                JSONObject contentJson = new JSONObject(content);
+                JSONObject resultJson = contentJson.getJSONObject("result");
+                JSONObject wattHourJson = resultJson.getJSONObject("watt_hours");
+                JSONObject wattJson = resultJson.getJSONObject("watts");
+                String zoneStr = contentJson.getJSONObject("message").getJSONObject("info").getString("timezone");
+                zone = ZoneId.of(zoneStr);
+                Iterator<String> iter = wattHourJson.keys();
+                // put all values of the current day into sorted tree map
+                while (iter.hasNext()) {
+                    String dateStr = iter.next();
+                    // convert date time into machine readable format
+                    try {
+                        ZonedDateTime zdt = LocalDateTime.parse(dateStr, dateInputFormatter).atZone(zone);
+                        wattHourMap.put(zdt, wattHourJson.getDouble(dateStr));
+                        wattMap.put(zdt, wattJson.getDouble(dateStr));
+                    } catch (DateTimeParseException dtpe) {
+                        logger.warn("Exception parsing time {} Reason: {}", dateStr, dtpe.getMessage());
+                        return;
+                    }
+                }
+                valid = true;
+            } catch (JSONException je) {
+                logger.debug("Error parsing JSON response {} - {}", content, je.getMessage());
+            }
+        }
+    }
+
+    public boolean isValid() {
+        if (valid) {
+            if (!wattHourMap.isEmpty()) {
+                if (expirationDateTime.isAfter(Instant.now())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public double getActualValue(ZonedDateTime queryDateTime) {
+        if (wattHourMap.isEmpty()) {
+            return UNDEF;
```

Comment:
This is a private constant returned by a public method, so the caller will only see this as `-1`. Since by contract you are returning -1 to signal "undefined", perhaps you should simply omit this constant which can also be confused by `UnDefType.UNDEF`. Javadoc might be more useful, WDYT?

Reply:
Ok for me. -1 is commonly used to indicate errors due to the fact all valid forecast values are positive. -1 is now directly used in code